### PR TITLE
Swerve into template

### PR DESCRIPTION
**Pull request by Rongrrz**

### CODE REVIEW

File: src/test/java/competition/subsystems/drive/DriveSubsystemTest.java [hunk 1]
```diff
@@ -9,10 +9,11 @@
 public class DriveSubsystemTest extends BaseCompetitionTest {
     @Test
     public void testTankDrive() {
```
Comment (stephenjust) [inline src/test/java/competition/subsystems/drive/DriveSubsystemTest.java]: Maybe just remove this test

File: src/main/java/competition/operator_interface/OperatorCommandMap.java [hunk 2]
```diff
@@ -18,9 +23,28 @@ public OperatorCommandMap() {}
...
     public void setupMyCommands(
             OperatorInterface operatorInterface,
-            SetRobotHeadingCommand resetHeading)
-    {
+            SetRobotHeadingCommand resetHeading,
+            DriveSubsystem drive) {
         resetHeading.setHeadingToApply(90);
         operatorInterface.gamepad.getifAvailable(1).onTrue(resetHeading);
+
+        // Stolen from 2024
```
Comment (stephenjust) [inline src/main/java/competition/operator_interface/OperatorCommandMap.java]: This looks game specific. Should we remove it?

**Review by stephenjust — CHANGES_REQUESTED** [review]

File: src/main/java/competition/operator_interface/OperatorCommandMap.java (hunk 2 above)
Comment (Rongrrz) [reply src/main/java/competition/operator_interface/OperatorCommandMap.java]: Will remove after testing the features :)

File: src/main/java/competition/electrical_contract/CompetitionContract.java [hunk 3]
```diff
@@ -16,12 +18,85 @@ public class CompetitionContract extends ElectricalContract {
...
+                    new DeviceInfo(getSteeringEncoderControllerName(swerveInstance), 53, false, simulationScalingValue);
+            case "RearRightDrive" ->
+                    new DeviceInfo(getSteeringEncoderControllerName(swerveInstance), 54, false, simulationScalingValue);
+            default -> null;
+        };
+    }
+
+    // Is this dependent on the robot?
+    @Override
+    public XYPair getSwerveModuleOffsets(SwerveInstance swerveInstance) {
```
Comment (JohnGilb) [inline src/main/java/competition/electrical_contract/CompetitionContract.java]: Follow up - let's label units

File: src/main/java/competition/operator_interface/OperatorCommandMap.java [hunk 4]
```diff
@@ -18,9 +23,9 @@ public OperatorCommandMap() {}
     @Inject
     public void setupMyCommands(
             OperatorInterface operatorInterface,
-            SetRobotHeadingCommand resetHeading)
-    {
-        resetHeading.setHeadingToApply(90);
+            SetRobotHeadingCommand resetHeading,
+            DriveSubsystem drive) {
```
Comment (JohnGilb) [inline src/main/java/competition/operator_interface/OperatorCommandMap.java]: Is this used for anything? Or can we remove it?

File: src/main/java/competition/subsystems/drive/SuggestedRotationValue.java [hunk 5]
```diff
@@ -0,0 +1,27 @@
+package competition.subsystems.drive;
```
Comment (JohnGilb) [inline src/main/java/competition/subsystems/drive/SuggestedRotationValue.java]: Not required for this PR, but this is a good candidate to move into the SCL.

File: src/main/java/competition/subsystems/drive/SuggestedRotationValue.java [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+    /**
+     * Used for SwerveDriveRotationAdvisor to pass back values in one of two ways:
+     * DesiredHeading (The heading you want to be at)
+     * HeadingPower (The power of which you should be rotating)
+     */
+    public double value;
+    public ValueType type;
+
+    public enum ValueType {
```
Comment (JohnGilb) [inline src/main/java/competition/subsystems/drive/SuggestedRotationValue.java]: ValueType isn't a very clear enum name - maybe `RotationGoalType`?

File: src/main/java/competition/subsystems/drive/SwerveDriveRotationAdvisor.java [hunk 7]
```diff
@@ -0,0 +1,140 @@
+package competition.subsystems.drive;
```
Comment (JohnGilb) [inline src/main/java/competition/subsystems/drive/SwerveDriveRotationAdvisor.java]: Not required for this PR, but this is a good candidate to move into the SCL.

File: src/main/java/competition/subsystems/drive/SwerveDriveRotationAdvisor.java [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+        // Don't think this takes in consideration of height...
+        Translation2d target = drive.getLookAtPointTarget();
+
+        Pose2d currentPose = pose.getCurrentPose2d();
+        aKitLogger.record("CurrentPoseX", currentPose.getX());
+        aKitLogger.record("CurrentPoseY", currentPose.getY());
+        Translation2d currentXY = new Translation2d(currentPose.getX(), currentPose.getY());
+
+        double desiredHeading = currentXY.minus(target).getAngle().getDegrees() + 180;
+        drive.setDesiredHeading(desiredHeading); // Does this need a recently heading reset check?
```
Comment (JohnGilb) [inline src/main/java/competition/subsystems/drive/SwerveDriveRotationAdvisor.java]: does it? :)

File: src/main/java/competition/subsystems/pose/PoseSubsystem.java [hunk 9]
```diff
@@ -30,4 +37,52 @@ protected double getRightDriveDistance() {
         return drive.getRightTotalDistance();
     }
 
+
+    private SwerveDrivePoseEstimator initializeSwerveOdometry() {
+        return new SwerveDrivePoseEstimator(
+                drive.getSwerveDriveKinematics(),
+                getCurrentHeading(),
+                new SwerveModulePosition[] {
```
Comment (JohnGilb) [inline src/main/java/competition/subsystems/pose/PoseSubsystem.java]: I think there's a method on the drive subsystem to get these all at once?

File: src/main/java/competition/subsystems/pose/PoseSubsystem.java [hunk 10]
```diff
@@ -30,4 +37,52 @@ protected double getRightDriveDistance() {
...
+        totalDistanceY = estimatedPosition.getY();
+
+        double prevTotalDistanceX = totalDistanceX;
+        double prevTotalDistanceY = totalDistanceY;
+        this.velocityX = ((totalDistanceX - prevTotalDistanceX));
+        this.velocityY = ((totalDistanceY - prevTotalDistanceY));
+        this.totalVelocity = (Math.sqrt(Math.pow(velocityX, 2.0) + Math.pow(velocityY, 2.0))); // Unnecessary?
+    }
+
+    private SwerveModulePosition[] getSwerveModulePositions() {
```
Comment (JohnGilb) [inline src/main/java/competition/subsystems/pose/PoseSubsystem.java:75]: or maybe use this method up on line 45

**Review by JohnGilb — CHANGES_REQUESTED** [review]

File: src/main/java/competition/subsystems/pose/PoseSubsystem.java (hunk 9 above)
Comment (Rongrrz) [reply src/main/java/competition/subsystems/pose/PoseSubsystem.java]: Unable to find, simplified using getSwerveModulePositions method instead.

**Review by JohnGilb — APPROVED** [review]

**Review by stephenjust — APPROVED** [review]

**Pull request merged**